packages/office-ui-fabric-react/src/components/Coachmark/Coachmark.tsx

```diff
@@ -326,8 +338,41 @@ export class Coachmark extends BaseComponent<ICoachmarkProps, ICoachmarkState> {
     );
   }
 
-  public componentWillUnmount(): void {
-    this._events.off(document, 'keydown', this._onKeyDown, true);
+  public dismiss = (ev?: Event | React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>): void => {
+    const { onDismiss } = this.props;
+
+    if (onDismiss) {
+      onDismiss(ev);
+    }
+  };
+
+  private _addListeners(): void {
+    const { preventDismissOnLostFocus } = this.props;
+    const currentDoc: Document = getDocument()!;
+
+    if (currentDoc) {
+      this._events.on(currentDoc, 'keydown', this._onKeyDown, true);
+
+      if (!preventDismissOnLostFocus) {
+        this._events.on(currentDoc, 'click', this._dismissOnLostFocus, true);
+        this._events.on(currentDoc, 'focus', this._dismissOnLostFocus, true);
+      } else {
+        this._events.off(currentDoc, 'click', this._dismissOnLostFocus, true);
```

Comment:
suggestion; don't call `off` anywhere but on the first line of this function. Calling `off` is totally fine, even if there are no events currently hooked up.

then you can just hook up the events.

Reply:
```tsx
private _addListeners(): void {
    const { preventDismissOnLostFocus } = this.props;
    const currentDoc: Document = getDocument()!;

  // Remove all events first, in case we call this method multiple times.
  this._events.off();

  // Hook up appropriate events.
  ...
}
```